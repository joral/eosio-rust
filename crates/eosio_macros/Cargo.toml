[package]
name = "eosio_macros"
version = "0.3.1"
authors = ["Liam Curry <liam@sagan.software>"]
license = "MIT OR Apache-2.0"
description = "Macros for creating compile-time EOSIO names and symbols."
edition = "2018"
keywords = ["eos", "eosio", "blockchain", "crypto", "dapp"]
categories = [
    "cryptography::cryptocurrencies",
    "encoding",
    "parsing",
    "value-formatting"
]
repository = "https://github.com/sagan-software/eosio-rust"
homepage = "https://sagan-software.github.io/eosio-rust/"
documentation = "https://sagan-software.github.io/eosio-rust/eosio-core-macros/"
readme = "README.md"
include = [
    "/Cargo.toml",
    "/src/**/*.rs",
    "/README.md",
    "/LICENSE-APACHE",
    "/LICENSE-MIT"
]

[badges]
maintenance = { status = "actively-developed" }
travis-ci = { repository = "sagan-software/eosio-rust", branch = "master" }

[dependencies]
eosio_macros_internal = { version = "0.3.1", path = "../eosio_macros_internal" }
proc-macro-hack = "0.5"

[dev-dependencies]
trybuild = "1.0"