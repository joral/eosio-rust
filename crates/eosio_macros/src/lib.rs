//! Macros for creating compile-time EOSIO names and symbols.
//!
//! Creating EOSIO names:
//!
//! ```
//! use eosio_macros::n;
//! assert_eq!(n!("test"), 14_605_613_396_213_628_928);
//! assert_eq!(n!("1234"), 614_248_767_926_829_056);
//! assert_eq!(n!("123451234512"), 614_251_535_012_020_768);
//! assert_eq!(n!("eosio.token"), 6_138_663_591_592_764_928);
//! ```
//!
//! Creating EOSIO symbols:
//!
//! ```
//! use eosio_macros::s;
//! assert_eq!(s!(4, "EOS"), 1397703940);
//! ```
#![no_std]
#![allow(clippy::missing_docs_in_private_items)]

use proc_macro_hack::proc_macro_hack;

/// Macro for converting EOSIO names into `u64` representations at compile
/// time.
///
/// # Examples
///
/// ```
/// use eosio_macros::n;
/// assert_eq!(n!("test"), 14_605_613_396_213_628_928);
/// assert_eq!(n!("1234"), 614_248_767_926_829_056);
/// assert_eq!(n!("123451234512"), 614_251_535_012_020_768);
/// assert_eq!(n!("eosio.token"), 6_138_663_591_592_764_928);
/// ```
#[proc_macro_hack]
pub use eosio_macros_internal::n;

/// Macro for converting EOSIO symbols into `u64` representations at
/// compile time.
///
/// # Examples
///
/// ```
/// use eosio_macros::s;
/// assert_eq!(s!(4, "EOS"), 1397703940);
/// ```
#[proc_macro_hack]
pub use eosio_macros_internal::s;

pub use eosio_macros_internal::{
    abi, action, table, NumBytes, Read, Table, Write,
};
