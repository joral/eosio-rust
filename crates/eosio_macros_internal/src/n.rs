use eosio_numstr::name_from_bytes;
use proc_macro2::{Literal, TokenStream};
use quote::{ToTokens, TokenStreamExt};
use syn::{
    parse::{Parse, ParseStream, Result as ParseResult},
    LitStr,
};

pub struct EosioName(u64);

impl Parse for EosioName {
    fn parse(input: ParseStream) -> ParseResult<Self> {
        let name = input.parse::<LitStr>()?.value();
        name_from_bytes(name.bytes())
            .map(Self)
            .map_err(|e| input.error(e))
    }
}

impl ToTokens for EosioName {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.append(Literal::u64_suffixed(self.0))
    }
}
