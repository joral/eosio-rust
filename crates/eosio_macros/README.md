[![Build Status](https://travis-ci.org/sagan-software/eosio-rust.svg?branch=master)](https://travis-ci.org/sagan-software/eosio-rust)

# eosio_macros

Macros for creating compile-time EOSIO names and symbols.

Creating EOSIO names:

```rust
use eosio_macros::n;
assert_eq!(n!("test"), 14_605_613_396_213_628_928);
assert_eq!(n!("1234"), 614_248_767_926_829_056);
assert_eq!(n!("123451234512"), 614_251_535_012_020_768);
assert_eq!(n!("eosio.token"), 6_138_663_591_592_764_928);
```

Creating EOSIO symbols:

```rust
use eosio_macros::s;
assert_eq!(s!(4, "EOS"), 1397703940);
```

License: MIT OR Apache-2.0
