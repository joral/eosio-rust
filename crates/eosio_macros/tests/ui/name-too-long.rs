extern crate eosio_macros;

use eosio_macros::n;

fn main() {
    let _ = n!("12345123451234");
}
