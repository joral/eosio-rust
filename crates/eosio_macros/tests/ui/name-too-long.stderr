error: unexpected end of input, name is too long, must be 13 chars or less
 --> $DIR/name-too-long.rs:6:13
  |
6 |     let _ = n!("12345123451234");
  |             ^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
